assets/helpers/paymentIntegrations/paymentApi.js

```diff
@@ -0,0 +1,240 @@
+import { type UsState, type CaState, type IsoCountry } from 'helpers/internationalisation/country';
+import { pollUntilP, logP } from 'helpers/promise';
+import trackConversion from 'helpers/tracking/conversions';
+
+import * as cookie from 'helpers/cookie';
+
+// ----- Types ----- //
+
+type OneOffFields = {|
+  paymentData: {
+    currency: IsoCurrency,
+    amount: number,
+    token: string,
+    email: string
+  },
+  acquisitionData: PaymentAPIAcquisitionData,
+|};
+
+type RegularContribution = {|
+  amount: number,
+  currency: string,
+  billingPeriod: BillingPeriod,
+|};
+
+type PayPalDetails = {| baid: string |};
+
+type StripeDetails = {| stripeToken: string |};
+
+type PaymentDetails = PayPalDetails | StripeDetails;
+
+type RegularFields = {|
+  firstName: string,
+  lastName: string,
+  country: IsoCountry,
+  state: UsState | CaState | null,
+  email: string,
+  contribution: RegularContribution,
+  paymentFields: PaymentDetails,
+  ophanIds: OphanIds,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  supportAbTests: AcquisitionABTest[],
+|};
+
+export type PaymentFields
+  = {| contributionType: 'oneoff', fields: OneOffFields |}
+  | {| contributionType: 'regular', fields: RegularFields |};
+
+type Credentials = 'omit' | 'same-origin' | 'include';
+
+export type Token
+  = {| paymentMethod: 'Stripe', token: string |}
+  | {| paymentMethod: 'PayPal', token: string |}
+  | {| paymentMethod: 'DirectDebit', accountHolderName: string, sortCode: string, accountNumber: string |};
+
+export type PaymentResult
+  = {| paymentStatus: 'success' |}
+  | {| paymentStatus: 'failure', error: string |};
+
+export type PaymentCallback = Token => Promise<PaymentResult>;
+
+// ----- Setup ----- //
+
+const PaymentSuccess: PaymentResult = { paymentStatus: 'success' };
+const POLLING_INTERVAL = 3000;
+const MAX_POLLS = 10;
+const ONEOFF_CONTRIB_ENDPOINT = window.guardian.paymentApiStripeEndpoint;
+
+// ----- Functions ----- //
+
+/** Builds a `RequestInit` object for use with GET requests using the Fetch API */
+function getRequestOptions(
+  credentials: Credentials,
+  csrf: CsrfState | null,
+): Object {
+  const headers = csrf !== null
+    ? { 'Content-Type': 'application/json', 'Csrf-Token': csrf.token || '' }
+    : { 'Content-Type': 'application/json' };
+
+  return {
+    method: 'GET',
+    headers,
+    credentials,
+  };
+}
+
+/** Builds a `RequestInit` object for use with POST requests using the Fetch API */
+function postRequestOptions(
+  data: PaymentFields,
+  credentials: Credentials,
+  csrf: CsrfState | null,
+): Object {
+  return { ...getRequestOptions(credentials, csrf), method: 'POST', body: JSON.stringify(data.fields) };
+}
+
+/** Sends a request to the payment API (or support workers, yolo) and converts
+ *  the response into a JSON object */
+function requestPaymentApi(endpoint: string, init: Object) {
+  return logP(fetch(endpoint, init).then(resp => resp.json()));
```

Comment:
Reading this through, I found `init: Object` slightly confusing. Maybe you could name `init` something more descriptive, or give it a tighter class.

Reply:
Also can we add a return type to this function?

Reply:
this function is actually so general I think it should have a more general name. `fetchJsonWithErrorLogging` or something like that? since it's completely independent of payment API, support workers, etc etc

Reply:
and it could also be moved somewhere under `assets/helpers`